### PR TITLE
common tests framework: cluster client creation could fail with invalid auth

### PR DESCRIPTION
**Pull request by chaochn47**
common tests framework: cluster client creation with auth could fail

related to #14041

> More than 2K line of code change is too big, could you try to break down this PR into small ones?

This PR has 227 additions and 73 deletions. 

Signed-off-by: Chao Chen <chaochn@amazon.com>

---

In order to fail the test cluster client creation with invalid auth, I have to change the interface function signature.

### CODE REVIEW

**Comment by chaochn47** [comment]
Does it look better or I can give another shot to split the PR? @ahrtr

File: tests/framework/e2e/etcdctl.go [hunk 1]
```diff
@@ -435,6 +459,20 @@ func (ctl *EtcdctlV3) AlarmDisarm(_ *clientv3.AlarmMember) (*clientv3.AlarmRespo
 	return &resp, err
 }
 
+func (ctl *EtcdctlV3) AuthEnable() (*clientv3.AuthEnableResponse, error) {
+	return &clientv3.AuthEnableResponse{}, SpawnWithExpect(ctl.cmdArgs("auth", "enable"), "Authentication Enabled")
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:463]: It doesn't make sense to return a empty `&clientv3.AuthEnableResponse{}` response. You have two choice:
1. Only return an error for the method;
2. Populate the `&clientv3.AuthEnableResponse{}` using `spawnJsonCmd`

File: tests/framework/config/cluster.go [hunk 2]
```diff
@@ -27,4 +27,6 @@ type ClusterConfig struct {
 	PeerTLS           TLSConfig
 	ClientTLS         TLSConfig
 	QuotaBackendBytes int64
+	AuthToken         string
+	SnapshotCount     uint64
```
Comment (ahrtr) [inline tests/framework/config/cluster.go:31]: `SnapshotCount` unrelated change?

File: tests/framework/config/cluster.go [hunk 3]
```diff
@@ -27,4 +27,6 @@ type ClusterConfig struct {
 	PeerTLS           TLSConfig
 	ClientTLS         TLSConfig
 	QuotaBackendBytes int64
+	AuthToken         string
```
Comment (ahrtr) [inline tests/framework/config/cluster.go:30]: I assume you will update some test cases to use this field in separate PR(s). Please let me know if it isn't the case.

Comment (chaochn47) [reply tests/framework/config/cluster.go:30]: yeah, this field will be used in real auth tests.

File: tests/framework/config/cluster.go (hunk 2 above)
Comment (chaochn47) [reply tests/framework/config/cluster.go:31]: This field will be used in real auth tests to trigger frequent snapshot to validate consistency.

File: tests/framework/e2e/etcdctl.go (hunk 1 above)
Comment (chaochn47) [reply tests/framework/e2e/etcdctl.go:463]: Good call. Will go with option 1

File: tests/framework/config/client.go [hunk 4]
```diff
@@ -20,6 +20,15 @@ import (
 	clientv3 "go.etcd.io/etcd/client/v3"
 )
 
+type ClientOption struct {
+	UserName string
+	Password string
+}
```
[the remark this answers is not in the document]
Comment (chaochn47) [reply tests/framework/config/client.go:26]: Nice. I am not aware of that. Will do

File: tests/framework/config/client.go [hunk 5]
```diff
@@ -20,6 +20,15 @@ import (
...
+
+func (opt ClientOption) Empty() bool {
+	return opt.UserName == "" && opt.Password == ""
+}
```
[the remark this answers is not in the document]
Comment (chaochn47) [reply tests/framework/config/client.go:30]: Will do

**Comment by chaochn47** [comment]
mixin test failure is unrelated. There is no monitoring changes. 

All the comments are addressed. PTAL, thx! @ahrtr

File: tests/framework/config/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahrtr) [reply tests/framework/config/client.go:30]: same as above

File: tests/framework/config/client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ahrtr) [reply tests/framework/config/client.go:26]: Why do you still keep `ClientOption`? My understanding is it should be removed, because it's duplicate to `AuthConfig`.

File: tests/framework/integration.go [hunk 6]
```diff
@@ -287,12 +308,28 @@ func (c integrationClient) LeaseRevoke(id clientv3.LeaseID) (*clientv3.LeaseRevo
 	return c.Client.Revoke(c.ctx, id)
 }
 
+func (c integrationClient) AuthEnable() (*clientv3.AuthEnableResponse, error) {
+	return c.Client.AuthEnable(context.Background())
```
Comment (serathius) [inline tests/framework/integration.go:312]:
```suggestion
	return c.Client.AuthEnable(c.ctx)
```

File: tests/framework/integration.go [hunk 7]
```diff
@@ -287,12 +308,28 @@ func (c integrationClient) LeaseRevoke(id clientv3.LeaseID) (*clientv3.LeaseRevo
...
+}
+
+func (c integrationClient) AuthDisable() (*clientv3.AuthDisableResponse, error) {
+	return c.Client.AuthDisable(context.Background())
```
Comment (serathius) [inline tests/framework/integration.go:316]: ditto

File: tests/framework/integration.go [hunk 8]
```diff
@@ -287,12 +308,28 @@ func (c integrationClient) LeaseRevoke(id clientv3.LeaseID) (*clientv3.LeaseRevo
...
+}
+
+func (c integrationClient) AuthStatus() (*clientv3.AuthStatusResponse, error) {
+	return c.Client.AuthStatus(context.Background())
```
Comment (serathius) [inline tests/framework/integration.go:320]: ditto

File: tests/framework/integration.go [hunk 9]
```diff
@@ -287,12 +308,28 @@ func (c integrationClient) LeaseRevoke(id clientv3.LeaseID) (*clientv3.LeaseRevo
...
+}
+
 func (c integrationClient) UserAdd(name, password string, opts config.UserAddOptions) (*clientv3.AuthUserAddResponse, error) {
 	return c.Client.UserAddWithOptions(c.ctx, name, password, &clientv3.UserAddOptions{
 		NoPassword: opts.NoPassword,
 	})
 }
 
+func (c integrationClient) UserGet(name string) (*clientv3.AuthUserGetResponse, error) {
+	return c.Client.UserGet(context.Background(), name)
```
Comment (serathius) [inline tests/framework/integration.go:330]: ditto

File: tests/framework/integration.go [hunk 10]
```diff
@@ -306,6 +343,14 @@ func (c integrationClient) UserChangePass(user, newPass string) error {
 	return err
 }
 
+func (c integrationClient) UserGrantRole(user string, role string) (*clientv3.AuthUserGrantRoleResponse, error) {
+	return c.Client.UserGrantRole(context.Background(), user, role)
```
Comment (serathius) [inline tests/framework/integration.go:347]: ditto

File: tests/framework/integration.go [hunk 11]
```diff
@@ -306,6 +343,14 @@ func (c integrationClient) UserChangePass(user, newPass string) error {
...
+}
+
+func (c integrationClient) UserRevokeRole(user string, role string) (*clientv3.AuthUserRevokeRoleResponse, error) {
+	return c.Client.UserRevokeRole(context.Background(), user, role)
```
Comment (serathius) [inline tests/framework/integration.go:351]: ditto

File: tests/framework/interface.go [hunk 12]
```diff
@@ -30,7 +30,8 @@ type testRunner interface {
 
 type Cluster interface {
 	Members() []Member
-	Client() Client
+	Client(cfg clientv3.AuthConfig) (Client, error)
+	MustClient(cfg clientv3.AuthConfig) Client
```
Comment (serathius) [inline tests/framework/interface.go:34]: There is no need to have a separate method `MustClient` for interface. We can keep the interface simple by having a normal function.

File: tests/framework/integration/cluster.go [hunk 13]
```diff
@@ -1410,32 +1410,41 @@ func (c *Cluster) Endpoints() []string {
...
-		if c.Cfg.ClientTLS != nil {
-			tls, err := c.Cfg.ClientTLS.ClientConfig()
-			if err != nil {
-				return nil, err
-			}
-			cfg.TLS = tls
-		}
-		c.clusterClient, err = newClientV3(cfg)
+func (c *Cluster) ClusterClient(opts ...func(*clientv3.Config)) (client *clientv3.Client, err error) {
+	if c.clusterClient != nil && len(opts) == 0 {
```
Comment (serathius) [inline tests/framework/integration/cluster.go:1414]: Adding ability to create a client from `opts` breaks assumptions needed to cache the client on `c.clusterClient`.

I think the original goal of caching client was removing need for user to call `.Close()` on client as this can be handled by Cluster.

If we allowing arguments for clients, I would prefer to remove caching altogether. Problem with closing client can be nicely done by adding it to `testing.T.Cleanup` as you nicely used in this PR above.

File: tests/framework/e2e/etcdctl.go [hunk 14]
```diff
@@ -39,6 +43,23 @@ func NewEtcdctl(cfg *EtcdProcessClusterConfig, endpoints []string) *EtcdctlV3 {
 	}
 }
 
+func (ctl *EtcdctlV3) WithAuth(userName, password string) (*EtcdctlV3, error) {
```
Comment (serathius) [inline tests/framework/e2e/etcdctl.go:46]: I would prefer to avoid mixing binary and library client in the same module. This hides fact how client interface is implemented, which is important for testing. We want to explicitly know that e2e tests use etcdctl and integration tests are using library. 

The tests you are working on are an exemption to the rule. They want to use client in e2e test. This is ok, however it should be explicit.

**Comment by chaochn47** [comment]
This PR is accidentally closed because I pushed to the same commit of main to the remote branch. How to re-open it?

**Comment by chaochn47** [comment]
This PR is ready to review. Would you mind take a look again? @ahrtr @serathius Thanks!!

File: tests/framework/config/client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (chaochn47) [reply tests/framework/config/client.go:26]: Will remove, thx!

File: tests/framework/integration.go (hunk 6 above)
Comment (chaochn47) [reply tests/framework/integration.go]: Done in the next PR revision

File: tests/framework/integration.go (hunk 7 above)
Comment (chaochn47) [reply tests/framework/integration.go]: Done in the next PR revision

File: tests/framework/integration.go (hunk 10 above)
Comment (chaochn47) [reply tests/framework/integration.go]: Done in the next PR revision

File: tests/framework/integration.go (hunk 11 above)
Comment (chaochn47) [reply tests/framework/integration.go]: Done in the next PR revision

File: tests/framework/interface.go (hunk 12 above)
Comment (chaochn47) [reply tests/framework/interface.go]: Done in the next PR revision

File: tests/framework/integration/cluster.go (hunk 13 above)
Comment (chaochn47) [reply tests/framework/integration/cluster.go]: Done in the next PR revision

File: tests/framework/e2e/etcdctl.go (hunk 14 above)
Comment (chaochn47) [reply tests/framework/e2e/etcdctl.go]: Done in the next PR revision

**Comment by chaochn47** [comment]
Ping again, do you think we still need this auth test migration? @serathius @ahrtr

File: tests/framework/integration.go [hunk 15]
```diff
@@ -261,17 +272,85 @@ func (c integrationClient) TimeToLive(ctx context.Context, id clientv3.LeaseID,
 	return c.Client.TimeToLive(ctx, id, leaseOpts...)
 }
 
+func (c integrationClient) LeaseList(ctx context.Context) (*clientv3.LeaseLeasesResponse, error) {
```
Comment (serathius) [inline tests/framework/integration.go:275]: Any reason you re-declare those functions?

File: tests/framework/integration.go [hunk 16]
```diff
@@ -315,6 +394,18 @@ func getOps(ss []string) ([]clientv3.Op, error) {
 	return ops, nil
 }
 
+func (c integrationClient) MemberAdd(ctx context.Context, _ string, peerAddrs []string) (*clientv3.MemberAddResponse, error) {
```
Comment (serathius) [inline tests/framework/integration.go:397]: Any reason you reorder the functions here? It only obscures git history.

File: tests/framework/integration/cluster.go [hunk 17]
```diff
@@ -178,8 +178,7 @@ type Cluster struct {
 	Members       []*Member
 	LastMemberNum int
 
-	mu            sync.Mutex
-	clusterClient *clientv3.Client
+	mu sync.Mutex
```
Comment (serathius) [inline tests/framework/integration/cluster.go:181]: Is the mutex still used? Can we remove it?

File: tests/framework/integration.go (hunk 15 above)
Comment (chaochn47) [reply tests/framework/integration.go:275]: It was added before Leases related PR is merged in upstream. I will rebase on latest code changes and send a new revision again. Thanks for pointing it out!!

File: tests/framework/integration.go (hunk 16 above)
Comment (chaochn47) [reply tests/framework/integration.go:397]: Oh, I wanted it to be in the same order of https://github.com/etcd-io/etcd/blob/cd9764a99fab81788d6234d3524a08c9b893c3e4/tests/framework/interface.go#L72-L79

File: tests/framework/integration/cluster.go [hunk 18]
```diff
@@ -1410,32 +1405,33 @@ func (c *Cluster) Endpoints() []string {
...
-		}
-		if c.Cfg.ClientTLS != nil {
-			tls, err := c.Cfg.ClientTLS.ClientConfig()
-			if err != nil {
-				return nil, err
-			}
-			cfg.TLS = tls
-		}
-		c.clusterClient, err = newClientV3(cfg)
+func (c *Cluster) ClusterClient(opts ...func(*clientv3.Config)) (client *clientv3.Client, err error) {
```
Comment (serathius) [inline tests/framework/integration/cluster.go:1408]: I don't think it's good to have inconsistency between `ClusterClient` and `Client` methods. Could update the integration. Cluster code so it also doesn't require user to call `Close`.

File: tests/framework/integration/cluster.go (hunk 17 above)
Comment (chaochn47) [reply tests/framework/integration/cluster.go:181]: It is stilled used when mutating `Members []*Member`. https://github.com/etcd-io/etcd/blob/cd9764a99fab81788d6234d3524a08c9b893c3e4/tests/framework/integration/cluster.go#L1371-L1375

Looks like I need to group it with `Members`, does it make sense to you?

File: tests/framework/integration/cluster.go (hunk 18 above)
Comment (chaochn47) [reply tests/framework/integration/cluster.go:1408]: In order to achieve that, I need to register the test clean up function whenever `ClusterClient` is called like this

```
	t.Cleanup(func() {
		cc.Close()
	})
```

However, `testing.TB` is not stored in the https://github.com/etcd-io/etcd/blob/cd9764a99fab81788d6234d3524a08c9b893c3e4/tests/framework/integration/cluster.go#L177-L184

Do you think one of the implementation can be adding this to the Cluster struct? Thanks!

**Comment by serathius** [comment]
For faster review I would recommend removing unrelated changes from the PR. It's over 300 lines of code, but still includes changes that muddy the water. Best if PRs do one thing and only this one thing.

**Comment by chaochn47** [comment]
> For faster review I would recommend removing unrelated changes from the PR. It's over 300 lines of code, but still includes changes that muddy the water. Best if PRs do one thing and only this one thing.

File: tests/framework/integration.go (hunk 16 above)
Comment (serathius) [reply tests/framework/integration.go:397]: Please separate logic changes from code cleanups, either separate commits or PRs.

File: tests/framework/integration/cluster.go (hunk 17 above)
Comment (serathius) [reply tests/framework/integration/cluster.go:181]: Hmm, usually the mutex declaration is above the variables it protects so it's a little unnatural. Let's keep it as it is and revisit it in future cleanup.

File: tests/framework/integration/cluster.go (hunk 18 above)
Comment (serathius) [reply tests/framework/integration/cluster.go:1408]: I think it's ok to add `testing.TB` to integration.Cluster.

File: tests/framework/integration.go (hunk 16 above)
Comment (chaochn47) [reply tests/framework/integration.go:397]: Make sense. Will do

File: tests/framework/e2e.go [hunk 19]
```diff
@@ -78,15 +82,31 @@ func (e e2eRunner) NewCluster(ctx context.Context, t testing.TB, cfg config.Clus
...
+		_, err := integration.NewClient(c.t, clientv3.Config{
+			Endpoints:   c.EndpointsV3(),
+			DialTimeout: 5 * time.Second,
+			DialOptions: []grpc.DialOption{grpc.WithBlock()},
+			Username:    cfg.Username,
+			Password:    cfg.Password,
+		})
+		if err != nil {
+			return nil, err
+		}
```
Comment (ahrtr) [inline tests/framework/e2e.go:106]: Probably we do not need this. Reasons:
1. It seems not good for e2e test to depends on integration test;
2. The related test cases will get a denied response if the auth info isn't correct. So the pre-check isn't necessary.

File: tests/framework/interface.go [hunk 20]
```diff
@@ -30,7 +30,7 @@ type testRunner interface {
 
 type Cluster interface {
 	Members() []Member
-	Client() Client
+	Client(cfg clientv3.AuthConfig) (Client, error)
```
Comment (ahrtr) [inline tests/framework/interface.go:33]: I am thinking whether it's possible to get rid of the parameter `cfg clientv3.AuthConfig`. When the test cases setup the user & role, and enable the auth, then the client should record the authConfig automatically. 

Does this make sense?

**Comment by ahrtr** [comment]
Please also rebase this PR although there is no conflict.

File: tests/framework/e2e.go (hunk 19 above)
Comment (chaochn47) [reply tests/framework/e2e.go:106]: I agree it's not optimal that e2e test depending on integration test. This is an exception when auth config is not empty. 

Copying from past conversation between @serathius and me. 

From me
> Some existing auth tests checks unauthenticated errors. 

> Integration test will throw an error at client creation stage with `clientv3.New` https://github.com/etcd-io/etcd/blob/7ec336f80c9463fecd6d8ef15a9f05f3e8f6287a/client/v3/client.go#L439

> However, e2e test won't fail at client creation stage and fail at "got response" stage instead. 

> Also every request from e2e test will establish a new grpc connection to the server but integration test reuses existing one. In order to use a different user name and password, integration test needs to create another client. 

From Marek
> I think we can create `clientv3.Client` just for this.

ref. https://github.com/etcd-io/etcd/pull/14041#discussion_r874629440

Does this makes sense to you?

File: tests/framework/interface.go (hunk 20 above)
Comment (chaochn47) [reply tests/framework/interface.go:33]: I don't think client should be stateful. In general, it makes the client not concurrent safe. If multiple callers set up the user & role and enable the auth using the client, this would cause contention.

Comment (ahrtr) [reply tests/framework/interface.go:33]: ack. 

It's just a little strange to provide a specific` cfg clientv3.AuthConfig` parameter, but automatically get all other parameters from the cluster.  Probably it makes more sense to resolve this using a generic way, such as change the parameter to something like `opts ...Option`.  It might be a little over engineering from another perspective. 

Let's keep it as it's for now, and probably think about it separately.

**Review by ahrtr — APPROVED** [review]
Overall looks good to me. 

Thank you @chaochn47

File: tests/framework/e2e.go (hunk 19 above)
Comment (serathius) [reply tests/framework/e2e.go:106]: > It seems not good for e2e test to depends on integration test;

It's not tests, it's e2e tests using integration client, which is just a wrapper for clientv3 library. There is nothing wrong with e2e tests using client library for some things that are not surfaced in etcdctl. To complete migration of tests to new framework this is even required as Some tests cannot be expressed using just e2e or integration tools

> I don't think e2e and integration test must follow the same pattern or behavior

But it would make testing and reasoning about them much simpler.

**Review by serathius — APPROVED** [review]
